### PR TITLE
ENH correctly restore default_factory of a defaultdict

### LINKED ISSUE #432: UntrustedTypesFoundException raised for standard library usage in lightgbm's Booster
Hello there,

after updating to 0.10.0 we got the following error:

```
...
File "/app/sales_prediction/common/result_storage.py", line 74, in get_train_result
return TrainResult(booster=skops.io.loads(model_data, trusted=["lightgbm.basic.Booster"]))
File "/opt/pysetup/.venv/lib/python3.10/site-packages/skops/io/_persist.py", line 191, in loads
audit_tree(tree)
File "/opt/pysetup/.venv/lib/python3.10/site-packages/skops/io/_audit.py", line 59, in audit_tree
raise UntrustedTypesFoundException(unsafe)
skops.io.exceptions.UntrustedTypesFoundException: Untrusted types found in the file: ['collections.defaultdict'].
```

In the serialised object we only have a single `lightgbm.basic.Booster` object.

1. Should the standard library be untrusted?
2. In this case the defaultdict is an internal details in the lightgbm's Booster. When the Booster is marked as trusted, should the internally used subdependencies still be untrusted?

### PR DESCRIPTION
**Pull request by adrinjalali**
Fixes #432 

During the inspection of #432 we also realized `default_factory` is not restored.

This also adds `defaultdict`, and a few other primitive builtin types as trusted.

cc @BenjaminBossan

### CODE REVIEW

File: skops/__init__.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 # Dev branch marker is: 'X.Y.dev' or 'X.Y.devN' where N is an integer.
 # 'X.Y.dev0' is the canonical version of 'X.Y.dev'
 #
-__version__ = "0.10.dev0"
+__version__ = "0.11.dev0"
```
Comment (BenjaminBossan) [inline skops/__init__.py:19]: Should this be in a separate PR?

File: skops/io/_trusted_types.py [hunk 2]
```diff
@@ -10,6 +10,10 @@
 
 PRIMITIVE_TYPE_NAMES = ["builtins." + t.__name__ for t in PRIMITIVES_TYPES]
 
+BUILTIN_TYPES = [list, set, map, tuple]
```
Comment (BenjaminBossan) [inline skops/io/_trusted_types.py:13]: I wonder about the name. Above we have "primitive types" but they are also "builtin". Should this be "complex types", "container types", "composite types" or something along these lines?

Then we could have `BUILTIN_TYPE_NAMES = PRIMITIVE_TYPE_NAMES + COMPLEX_TYPE_NAMES` (or whatever name is chosen) and save a bit of code duplication.

File: skops/io/tests/test_persist.py [hunk 3]
```diff
@@ -1078,3 +1081,12 @@ def test_trusted_bool_raises(tmp_path):
...
         loads(dumps(10), trusted=True)  # type: ignore
+
+
+def test_defaultdict():
+    """Test that we correctly restore a defaultdict."""
+    obj = defaultdict(set)
+    obj["foo"] = "bar"
+    obj_loaded = loads(dumps(obj))
+    assert obj_loaded == obj
+    assert obj_loaded.default_factory == obj.default_factory
```
Comment (BenjaminBossan) [inline skops/io/tests/test_persist.py:1092]: Should a test for `OrderedDict` be added too?

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for fixing this bug. I added a few comments, please take a look.

**Comment by adrinjalali** [comment]
@BenjaminBossan thanks for the review, I think I addressed your comments

File: skops/io/tests/test_persist.py [hunk 4]
```diff
@@ -1092,3 +1092,10 @@ def test_defaultdict():
     obj_loaded = loads(dumps(obj))
     assert obj_loaded == obj
     assert obj_loaded.default_factory == obj.default_factory
+
+
+@pytest.mark.parametrize("cls", [dict, OrderedDict])
+def test_dictionary(cls):
+    obj = cls({1: 5, 6: 3, 2: 4})
+    loaded_obj = loads(dumps(obj))
+    assert obj == loaded_obj
```
Comment (BenjaminBossan) [inline skops/io/tests/test_persist.py:1101]: Should the type also be checked? E.g. this passes:

```python
assert {2: 20, 1: 10} == OrderedDict([(2, 20), (1, 10)])
assert {1: 10, 2: 20} == OrderedDict([(2, 20), (1, 10)])
```

Comment (BenjaminBossan) [reply skops/io/tests/test_persist.py:1101]: ping @adrinjalali

Comment (adrinjalali) [reply skops/io/tests/test_persist.py:1101]: oops, sorry. Added now @BenjaminBossan

**Review by BenjaminBossan — APPROVED** [review]
Thanks for fixing this, LGTM.

Failing CI seems to be unrelated, some tests fail because of a change in sklearn and some because MacOS is no longer supported by codecov (??). So feel free to merge despite the red CI.

**Comment by juhoinkinen** [comment]
Do you have an ETA for a release which would include this fix?

**Comment by adrinjalali** [comment]
Working on some sklearn 1.6 compatibility issues (due to come out soon) and then we'd do a release.
